\documentclass{bioinfo}
\copyrightyear{2005}
\pubyear{2005}
\begin{document}
\firstpage{1}

\title[SPARQL-enabled identifier conversion with Identifiers.org]{SPARQL-enabled identifier conversion with Identifiers.org}
\author[Sarala M. Wimalaratne \textit{et~al}]{Sarala M Wimalaratne\,$^{1, \footnote{To whom correspondence should be addressed}, \footnote{The first two authors should be regarded as joint first authors}}$, Jerven Bolleman\,$^{2,  \footnotemark[\value{footnote}]}$, Nick Juty\,$^{1}$, Toshiaki Katayama\,$^{3}$, Michel Dumontier\,$^{4}$, Nicole Redaschi\,$^{2}$, Nicolas Le Nov{\`e}re\,$^{1,5}$, Henning Hermjakob\,$^{1}$, and Camille Laibe\,$^1$}
\address{$^{1}$European Molecular Biology Laboratory, European Bioinformatics Institute (EMBL-EBI), Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK \\
$^{2}$Swiss-Prot group, SIB Swiss Institute of Bioinformatics, Centre Medical Universitaire, 1211 Geneve, Switzerland \\
$^{3}$Database Center for Life Science, Research Organization of Information and Systems, 178-4-4 Wakashiba, Kashiwa, Chiba 277-0871, Japan \\
$^{4}$Stanford Center for Biomedical Informatics Research, Stanford University, CA 94305-5479, USA \\
$^{5}$Babraham Institute, Babraham Research Campus, Cambridge, CB22 3AT, UK}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

\maketitle

\begin{abstract}
\section{Motivation:}
On the Semantic Web, in life sciences in particular, data is often distributed via multiple resources. Each of these sources is likely to use their own IRI (International Resource Identifier) for conceptually the same resource or database record. The lack of correspondence between identifiers introduces a barrier when executing federated SPARQL queries across life science data.

\section{Results:}
We introduce a novel SPARQL-based service to enable on-the-fly integration of life science data. This service uses the identifiers patterns defined in the Identifiers.org registry to generate a plurality of identifier variants, which can then be used to match source identifiers with target identifiers. We demonstrate the utility of this identifier integration approach by answering queries across major producers of life science Linked Data.

\section{Availability:}
The SPARQL-based identifier conversion service is available without restriction at \linebreak\href{http://identifiers.org/services/sparql}{http://identifiers.org/services/sparql}.

\section{Contact:} \href{sarala@ebi.ac.uk}{Sarala M Wimalaratne}
\end{abstract}

\section{Introduction}
Semantic Web technologies such as the Resource Description Framework (RDF; http://www.w3.org/TR/rdf-primer/) offer a powerful paradigm for publishing and exploring life science data through standardization of format and data access. For example, the open source Bio2RDF (\cite{Biordf2}) projects converts dozens of public biological databases and datasets from legacy formats into RDF, and provides a mechanism to explore these as Linked Data. Recently, established bioinformatic organizations such as DBCLS (\cite{Kawano2014}), NCBI (https://pubchem.ncbi.nlm.nih.gov/rdf/), NexProt (\cite{Chichester2014}), and the EMBL-EBI in collaboration with the UniProt consortium have made some datasets available in RDF (\cite{Jupp01052014}), thereby significantly extending the network of the Linked Open Data .

Each network uses HTTP-based International Resource Identifiers (IRIs) to identify and link data items. This facilitates querying across network-linked resources, but the lack of a universal identifier system requires mappings across all the different identifiers in use. Identifiers.org (\cite{Juty01012012}) provides resolvable persistent IRIs used to identify individual records (based on the existing entity identifiers assigned directly by the data providers). Although some linked data providers such as Bio2RDF and the EBI now make their data available with identifiers.org URIs (or mappings to them), this practice is not widely implemented. Therefore, the identifier mismatch makes it difficult to query multiple datasets simultaneously.  String manipulation, supported by SPARQL, may be used for this purpose but requires users to know in advance the IRI types being used in each resource, making it a cumbersome and inefficient solution. 
 
To address the issue of identifier heterogeneity, we have developed a SPARQL-based service that generates on-the-fly identifier mappings for registered IRI patterns. Here, we describe our novel method and demonstrate its functionality through service-enabled federated SPARQL queries. This system offers an automatic way to link and query over a rapidly growing number of semantic web friendly life science datasets.

\begin{methods}
\section{Methods}
We implemented a SPARQL-based service that generates a set of variant identifiers based on a provided identifier. This service, implemented using the OpenRDF Sesame SPARQL engine (http://www.openrdf.org/), translates an incoming query pattern of the form <subjectIRI> $\mathtt{owl\textcolon{}sameAs}$ ?targetIRI and generates a set of triples with the specific subject, predicate, and the generated target IRI. The service queries the curated Identifiers.org Registry to determine the originating data collection, then obtains alternative IRIs patterns, and finally generates and returns alternative IRIs. While this implementation queries the relational database that underpins the Identifiers.org Registry, others could implement a similar service using the Registry's public web services.
\end{methods}

\section{Results}
The Identifiers.org SPARQL service can be used to list all alternative IRI schemes available for an IRI pattern that is available in the Identifiers.org Registry. Currently, the Identifiers.org Registry contains 531 datasets and 1332 IRI patterns, therefore providing a total of 707292 IRI translations. For supported datasets, this service eliminates the need to know the set of valid IRI patterns in advance and the need to devise elaborate string manipulation operations in a federated SPARQL query (figure \ref{example1}).

The query below illustrates how the service can be used to query across datasets with different IRI schemes. In this example, we obtain the definitions for Gene Ontology terms that are used to annotate the species in a model from the BioModels database. The service bridges the gap between the Identifiers.org-specified and the Bio2RDF-specified identifier for the Gene Ontology term. This query can be executed from the BioModels' SPARQL endpoint.

\begin{figure}[h]
%PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
%PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
%PREFIX owl: <http://www.w3.org/2002/07/owl#>
%PREFIX dcterms: <http://purl.org/dc/terms/>
%PREFIX sbmlrdf: <http://identifiers.org/biomodels.vocabulary#>
%PREFIX bqbio: <http://biomodels.net/biology-qualifiers#>
{\scriptsize
\begin{verbatim}
SELECT DISTINCT ?species ?go_term ?go_description WHERE {
   ?model sbmlrdf:species ?s . 
   ?s sbmlrdf:name ?species .
   ?s bqbio:isVersionOf ?go_term 
   FILTER (?model = <http://identifiers.org/biomodels.db/BIOMD0000000001>)
   FILTER regex(?go_term,"go","i")
   SERVICE <http://identifiers.org/services/sparql>{
      ?go_term owl:sameAs ?go .
   }
   SERVICE <http://bioportal.bio2rdf.org/sparql>{
      ?go dcterms:description ?go_description .
   }
}LIMIT 10
\end{verbatim}
}
\caption{Using Identifiers.org SPARQL service to transform Identifiers.org IRIs into Bio2RDF IRIs.}
\label{example1}
\end{figure}

Similarly we could write a SPARQL query at the uniprot.org SPARQL endpoint that combines information in UniProt and Wikipathways. SERVICE clause using Identifiers.org SPARQL endpoint to transform the purl.uniprot.org form of IRIs to Identifiers.org which are used in UniProtKB and Wikipathways, respectively (figure \ref{example2}). Further examples including the queries discussed here are available at \href{http://identifiers.org/documentation}{http://identifiers.org/documentation}.

\begin{figure}[h]
%PREFIX up: <http://purl.uniprot.org/core/>
%PREFIX taxon: <http://purl.uniprot.org/taxonomy/>
%PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
%PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
%PREFIX faldo: <http://biohackathon.org/resource/faldo#>
%PREFIX owl: <http://www.w3.org/2002/07/owl#>
%PREFIX dcterms: <http://purl.org/dc/terms/>
%PREFIX dc: <http://purl.org/dc/elements/1.1/>
{\scriptsize
\begin{verbatim}		
SELECT ?protein ?diseaseComment ?pathway ?pathwayName WHERE {
   ?protein up:annotation/up:disease/rdfs:comment ?diseaseComment .
   SERVICE <http://identifiers.org/services/sparql>{
      ?protein owl:sameAs ?otherIRIs .
   }
   SERVICE <http://sparql.wikipathways.org/> {
      ?x dcterms:isPartOf ?pathway; ?z ?otherIRIs .
      ?pathway dc:title ?pathwayName
   }
}
LIMIT 10
\end{verbatim}
}
\caption{Using Identifiers.org SPARQL service to transform UniProt IRIs into Identifiers.org IRIs.}
\label{example2}
\end{figure}
    
\section{Discussion}
Leveraging the wealth of biomedical big data for discovery requires simple and effective approaches to tame the challenge of working with heterogeneous, overlapping, and diverse data. Of particular concern is assignment of different identifiers for identical resources as well as for conceptually identical resources. Identifier integration is the subject of much research that focuses either on integrating conceptually identical objects or their relations (\cite{VanIersel2010}, \cite{Wein2012}, \cite{Smith2007}, \cite{Chambers2013}). In contrast, our work focuses on the problem of having multiple identifiers for the same database object, which is an emerging issue among semantic web data providers. Our solution is rapid, scalable, and will grow to provide new identifier-based mappings as additional IRI patterns are added to the Identifiers.org Registry. 

\section{Conclusion}
This IRI conversion service, provided by Identifiers.org as a SPARQL service, will enable users to focus on asking meaningful questions across biological datasets of interest rather than figuring out how to generate the right identifiers. 

\section*{Acknowledgement}
The authors wish to thank the DBCLS RDF Summit and Biohackathon organisers for fostering the initial discussions and development efforts, and Simon Jupp for developing the generic LODEStar user interface used by this endpoint.

\paragraph{Funding\textcolon} 
This work received support from the BBSRC (BB/J019305/1), Japanese grants (?), National Institutes of Health grants (1U41HG006104), the Swiss Federal Government through the Federal Office of Education and Science, and EMBL.  

\bibliographystyle{natbib}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%\bibliographystyle{bioinformatics}
%\bibliographystyle{plain}
%\bibliography{Document}
\bibliography{document}  


\end{document}
