\documentclass{bioinfo}
\copyrightyear{2005}
\pubyear{2005}
\begin{document}
\firstpage{1}

\title[Identifiers.org virtual SPARQL endpoint for IRI schemes conversion]{Identifiers.org virtual SPARQL endpoint for IRI schemes conversion}
\author[Sarala M. Wimalaratne \textit{et~al}]{Sarala M Wimalaratne\,$^{1, \footnote{To whom correspondence should be addressed}, \footnote{The first two authors should be regarded as joint first authors}}$, Jerven Bolleman\,$^{2,  \footnotemark[\value{footnote}]}$, Nick Juty\,$^{1}$, Toshiaki Katayama\,$^{3}$, Michel Dumontier\,$^{4}$, Nicole Redaschi\,$^{2}$, Nicolas Le Nov{\`e}re\,$^{5}$, Henning Hermjakob\,$^{1}$, and Camille Laibe\,$^1$}
\address{$^{1}$European Molecular Biology Laboratory, European Bioinformatics Institute (EMBL-EBI), Wellcome Trust Genome Campus, Hinxton, Cambridge CB10 1SD, UK \\
$^{2}$Swiss-Prot group, SIB Swiss Institute of Bioinformatics, Centre Medical Universitaire, 1211 Geneve, Switzerland \\
$^{3}$Database Center for Life Science, Research Organization of Information and Systems, 178-4-4 Wakashiba, Kashiwa, Chiba 277-0871, Japan \\
$^{4}$Stanford Center for Biomedical Informatics Research, Stanford University, CA 94305-5479, USA \\
$^{5}$Babraham Institute, Babraham Research Campus, Cambridge, CB22 3AT, UK}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

\maketitle

\begin{abstract}
\section{Motivation:}
On the Semantic Web, in life sciences in particular, data is often distributed via multiple resources. Each of these sources is likely to use their own IRI (International Resource Identifier) for conceptually the same resource or database record. This multitude of identifiers introduces a barrier when executing federated SPARQL queries between different datasets.

\section{Results:}
We introduce a virtual SPARQL endpoint provided by Identifiers.org, using its Registry's knowledge about IRI patterns, to automatically translate identifiers from one IRI pattern to another, making querying across resources easier and more robust.

\section{Availability:}
This IRI translating endpoint is available at \linebreak\href{http://identifiers.org/services/sparql}{http://identifiers.org/services/sparql} and is free to use for all.

\section{Contact:} \href{sarala@ebi.ac.uk}{Sarala M Wimalaratne}
\end{abstract}

\section{Introduction}
The use of Semantic Web technologies such as the Resource Description Framework (RDF; \cite{Manola2004}) for publishing data is becoming popular in the life sciences.
For example, the Bio2RDF (\cite{biordf2}) data mashup is one such resource built upon the data from various publicly available biological databases translated into RDF and linked together. More recently, EMBL-EBI in collaboration with the UniProt consortium has exposed some datasets in RDF via the EBI RDF platform (\cite{Jupp01052014}).

Bio2RDF uses normalized International Resource Identifier (IRI) into its own domain name to support queries across datasets.
Datasets that are part of the EBI RDF platform follow a set of rules regarding the IRIs they can use, resulting in a consistent set of IRIs used across all participating resources. Hence, while these two platforms provide linked datasets which can be easily queried individually, the use of two different IRI schemes for what are essentially the same concepts makes it difficult to perform cross platform queries using SPARQL (\cite{Prud'hommeaux2008}).  

SPARQL, a query language for RDF which supports complex queries, can be used to merge data distributed over multiple RDF resources. To enable querying across different platforms, there is a need to rewrite IRIs on the fly. 
String manipulation, supported by SPARQL, may be used for this purpose but requires users to know in advance the IRI types being used in each resource, making it a cumbersome and inefficient solution. 

Identifiers.org (\cite{Juty01012012}) provides resolvable persistent IRIs used to identify individual records (based on the existing entity identifiers assigned directly by the data providers). IRI assignment is reliant upon an underlying Registry which contains information about individual datasets (data collection), the resources (or physical locations) which distribute them, and the different IRIs used to identify and access the individual records.

To resolve cross query issues, Identifiers.org has developed a virtual SPARQL endpoint to support IRI conversion between linked datasets using different IRI schemes. Consequently, this provides an automatic system for linking and querying ever growing number of life science data resources.

\begin{methods}
\section{Methods}
The OpenRDF Sesame SPARQL engine (\cite{SESAME}) was extended to translate specific query patterns to allow IRI translation. Specifically, the service has been designed to answer queries requesting IRIs and making use of the predicate $\mathtt{owl\textcolon{}sameAs}$. When such queries are detected, the system determines from which data collection the identified entity originates, extracts from the Registry the alternative root IRIs for this concept, regenerates complete IRI alternative forms and returns them. The main integration point within the SESAME engine is the $getStatements$ method of a class extending $org.openrdf.query.algebra.evaluation.TripleSource$.

The database underlying Identifiers.org's Registry was extended in order to record alternative IRI schemes. This has additionally required a curational effort to ensure the necessary information has been stored in the Registry.

This is a ``virtual'' SPARQL endpoint, directly driven by the relational database underpinning the Registry.
While the same extension logic could be coded directly in a native SPARQL platform. However we chose to reuse the existing Identifiers.org infrastructure for pragmatic reasons. 
LODEStar (\cite{LODEStar}), used by the EBI RDF platform, has been deployed to provide a query interface for users.
\end{methods}

\section{Results}
The Identifiers.org virtual SPARQL endpoint can be used to list all alternative IRI schemes available for a given IRI. It can also be used to check whether two IRIs describes the same concept (although it is not a mapping service between entities provided by different datasets). Example queries can be found at the endpoint page URL.

With the Identifiers.org SPARQL service, it is possible to avoid using string manipulation within a SPARQL query (figure \ref{translatestuff}) and does not require users to know which IRI patterns are being used within a given resource.

The query below illustrates how the service can be used to integrate resources using different IRI schemes. 
In this example, the resources are the BioModels Linked Dataset and Bio2RDF which uses IRIs of the form \href{http://identifiers.org/go/GO:0006915}{http://identifiers.org/go/GO:0006915} and \href{http://bio2rdf.org/go:0006915} {http://bio2rdf.org/go:0006915} for Gene Ontology terms (\cite{Ashburner2000}). 
The query lists the cross-references to GO terms present in the computational model ``Edelstein1996 - EPSP ACh event'' (BIOMD0000000001).
It is executed from the BioModels' SPARQL endpoint (\href{http://www.ebi.ac.uk/rdf/services/biomodels/sparql}{http://www.ebi.ac.uk/rdf/services/biomodels/sparql}) and connects to Identifiers.org to retrieve the alternative IRIs for the retrieved GO terms. Those are then passed to the Bio2RDF SPARQL endpoint to retrieve the description of each GO term.

\begin{figure}[h]
%PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
%PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
%PREFIX owl: <http://www.w3.org/2002/07/owl#>
%PREFIX dc: <http://purl.org/dc/elements/1.1/>
%PREFIX dcterms: <http://purl.org/dc/terms/>
%PREFIX sbmlrdf: <http://identifiers.org/biomodels.vocabulary#>
%PREFIX bqbio: <http://biomodels.net/biology-qualifiers#>
{\scriptsize
\begin{verbatim}
SELECT DISTINCT ?species ?annotation ?description WHERE { 
   <http://identifiers.org/biomodels.db/BIOMD0000000001> 
      sbmlrdf:species ?species . 
   ?species bqbio:isVersionOf ?annotation .
   
   SERVICE <http://identifiers.org/services/sparql>{
      ?annotation owl:sameAs ?otherIRIs .
   }
   
   SERVICE <http://bioportal.bio2rdf.org/sparql>{
      ?otherIRIs dcterms:description ?description .
   }        
}LIMIT 10
\end{verbatim}
}
  \caption{Using Identifiers.org virtual SPARQL endpoint to transform Identifiers.org IRIs into Bio2RDF IRIs.}
  \label{translatestuff}
\end{figure}

\section{Discussion}
There are a number of mapping services aimed to connect identifiers from one database to another. However, these are conceptual mappings, for instance relating a gene identifier to its protein product. In contrast, Identifiers.org records all access URLs and IRI-based identification schemes used to identify individual entities from a single data collection.

Also, while the produced endpoint is SPARQL 1.1 compliant, its virtual nature (not based on an actual triple store) means the queries available are limited to IRIs using the $\mathtt{owl\textcolon{}sameAs}$ predicate in a basic graph pattern; otherwise no results will be returned. The correct SPARQL solution would be to return the infinite possible IRI translations, but although a technically correct solution, would not be useful in practice. Instead we take the approach that Identifiers.org SPARQL endpoint is a remote service not under direct control by the user, and in this case it is under control of entity that regenerate a new RDF graph on the fly for each query.

This solution provides users the illusion of a valid SPARQL service that can translate all IRI patterns recorded in the Identifier.org Registry as if all these queries where materialised, at a fraction of the cost.

\section{Conclusion}
This IRI conversion service, provided by Identifiers.org as a virtual SPARQL endpoint, enhances the federated query capacities of SPARQL 1.1 to effectively combine information from multiple heterogeneous datasets.

\section*{Acknowledgement}
The authors wish to thank the DBCLS RDF Summit and Biohackathon organisers for fostering the initial discussions and development efforts, and Simon Jupp for developing the generic LODEStar user interface used by this endpoint.

\paragraph{Funding\textcolon} 
This work received support from the BBSRC (BB/J019305/1), Japanese grants (?), National Institutes of Health grants (1U41HG006104), the Swiss Federal Government through the Federal Office of Education and Science, and EMBL.  

\bibliographystyle{natbib}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%\bibliographystyle{bioinformatics}
%
%\bibliographystyle{plain}
%
%\bibliography{Document}
\bibliography{document}  


\end{document}
